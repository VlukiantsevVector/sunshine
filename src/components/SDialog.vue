<template>
  <div
    v-if="visible"
    :class="classes('wrapper')"
    @click.self="close"
  >
    <div :class="classes({ visible })">
      <div :class="classes('header')">
        <span :class="classes('heading')">{{ heading }}</span>
        <s-icon @click="close" package="sunshine24" name="close" />
      </div>
      <form v-bind="$attrs" v-on="$listeners">
        <slot v-bind="{ close }" />
      </form>
    </div>
  </div>
</template>

<script>
import Vue from 'vue';
import { ClassesMixin } from '../lib/utils';

export default Vue.extend({
  name: 's-dialog',
  inheritAttrs: false,

  mixins: [
    ClassesMixin,
  ],

  model: {
    prop: 'visible',
    event: 'toggle',
  },

  props: {
    visible: {
      type: Boolean,
      default: true,
    },

    heading: String,
  },

  methods: {
    close() {
      this.$emit('toggle', false);
    },
  },

  watch: {
    // HACK
    async visible(next, prev) {
      if (next && !prev) {
        await Vue.nextTick();
        this.$el.querySelector('input').focus();
      }
    },
  },
});
</script>
