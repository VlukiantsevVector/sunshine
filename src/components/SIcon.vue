<template>
  <span
    :class="classes()"
    v-on="$listeners"
  ><span
    :style="iconStyle"
    :class="iconClass"
  />
  </span>
</template>

<script lang="ts">
import Vue from 'vue';
import mixins from 'vue-typed-mixins';
import ClassesMixin from './internal/ClassesMixin';

import '../icons/icon54/style.css';
import '../icons/sunshine24/css/sunshine24-embedded.css';


export default mixins(ClassesMixin).extend({
  name: 's-icon',

  props: {
    package: {
      type: String,
      default: 'icon',
    },

    name: {
      type: String,
      required: true,
    },

    rotate: {
      type: Number,
      default: 0,
    },
  },

  computed: {
    iconClass(): string {
      return `${this.package}-${this.name}`;
    },

    iconStyle(): any {
      return {
        display: 'inline-block',
        transform: `rotate(${this.rotate}deg)`,
      };
    },
  },
});
</script>
