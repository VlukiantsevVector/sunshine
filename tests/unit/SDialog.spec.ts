import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import SDialog from '@/components/SDialog.vue';
import SIcon from '@/components/SIcon.vue';

const localVue = createLocalVue();
localVue.component('SIcon', SIcon);

describe('SDialog', () => {
  it('is hidden', () => {
    const wrapper = shallowMount(SDialog, {
      localVue,
      propsData: {
        visible: false,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('is visible', () => {
    const wrapper = shallowMount(SDialog, {
      localVue,
      propsData: {
        visible: true,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays heading', () => {
    const wrapper = shallowMount(SDialog, {
      localVue,
      propsData: {
        visible: true,
        heading: 'header',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('displays slot content', () => {
    const wrapper = shallowMount(SDialog, {
      localVue,
      propsData: {
        visible: true,
      },
      scopedSlots: {
        default: '<button>OK</button>',
      },
    });

    expect(wrapper.contains('button')).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('closes on pressing SIcon', () => {
    const wrapper = mount(SDialog, {
      components: {
        SIcon,
      },
      propsData: {
        visible: true,
      },
    });

    let sIcon = wrapper.find('.sunshine24-close');
    sIcon.trigger('click');

    expect(wrapper.emitted('toggle')[0][0]).toEqual(false);
  });

  // test: 'closes on close button passed by slot'?
});
