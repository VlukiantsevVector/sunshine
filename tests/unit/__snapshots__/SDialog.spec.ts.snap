// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SDialog displays heading 1`] = `
<div class="s-dialog__wrapper">
  <div class="s-dialog s-dialog--visible">
    <div class="s-dialog__header"><span class="s-dialog__heading">header</span>
      <s-icon-stub package="sunshine24" name="close" rotate="0"></s-icon-stub>
    </div>
    <form></form>
  </div>
</div>
`;

exports[`SDialog displays slot content 1`] = `
<div class="s-dialog__wrapper">
  <div class="s-dialog s-dialog--visible">
    <div class="s-dialog__header"><span class="s-dialog__heading"></span>
      <s-icon-stub package="sunshine24" name="close" rotate="0"></s-icon-stub>
    </div>
    <form><button>OK</button></form>
  </div>
</div>
`;

exports[`SDialog is hidden 1`] = `undefined`;

exports[`SDialog is visible 1`] = `
<div class="s-dialog__wrapper">
  <div class="s-dialog s-dialog--visible">
    <div class="s-dialog__header"><span class="s-dialog__heading"></span>
      <s-icon-stub package="sunshine24" name="close" rotate="0"></s-icon-stub>
    </div>
    <form></form>
  </div>
</div>
`;
